tests/benchmark/stateful/bloatnet/test_single_opcode.py

```diff
@@ -0,0 +1,177 @@
+ALLOWANCE_SELECTOR = 0xDD62ED3E  # allowance(address,address)
+
+
+# SLOAD BENCHMARK ARCHITECTURE:
+#
+#   [Pre-deployed ERC20 Contract] ──── Storage slots for balances
+#           │
+#           │  balanceOf(address) → SLOAD(keccak256(address || slot))
+#           │
+#   [Attack Contract] ──CALL──► ERC20.balanceOf(random_address)
+#           │
+#           └─► Loop(i=0 to N):
+#                 1. Generate random address from counter
+#                 2. CALL balanceOf(random_address) → forces cold SLOAD
+#                 3. Most addresses have zero balance → empty storage slots
+#
+# WHY IT STRESSES CLIENTS:
+#   - Each balanceOf() call forces a cold SLOAD on a likely-empty slot
+#   - Storage slot = keccak256(address || balances_slot)
+#   - Random addresses ensure maximum cache misses
+#   - Tests client's sparse storage handling efficiency
+
+
+@pytest.mark.valid_from("Prague")
+def test_sload_empty_erc20_balanceof(
+    blockchain_test: BlockchainTestFiller,
+    pre: Alloc,
+    fork: Fork,
+    gas_benchmark_value: int,
+    address_stubs,
+):
+    """
+    BloatNet SLOAD benchmark using ERC20 balanceOf queries on random addresses.
+
+    This test:
+    1. Auto-discovers ERC20 contracts from stubs (pattern: erc20_contract_*)
+    2. Splits gas budget evenly across all discovered contracts
+    3. Queries balanceOf() incrementally starting by 0 and increasing by 1.
+    (thus forcing SLOADs to non-existing addresses.)
+    """
+    gas_costs = fork.gas_costs()
+
+    # Calculate gas costs
+    intrinsic_gas = fork.transaction_intrinsic_cost_calculator()(calldata=b"")
+
+    # Cost per iteration (attack contract overhead + balanceOf call)
+    cost_per_iteration = (
+        # Attack contract loop overhead
+        gas_costs.G_VERY_LOW * 2  # MLOAD counter (3*2)
+        + gas_costs.G_VERY_LOW * 2  # MSTORE selector (3*2)
+        + gas_costs.G_VERY_LOW * 3  # MLOAD + MSTORE address (3*3)
+        + gas_costs.G_BASE  # POP (2)
+        + gas_costs.G_BASE * 3  # SUB + MLOAD + MSTORE for counter decrement (2*3)
+        + gas_costs.G_BASE * 2  # ISZERO * 2 for loop condition (2*2)
+        + gas_costs.G_MID  # JUMPI (8)
+        # CALL to ERC20 contract
+        + gas_costs.G_WARM_ACCOUNT_ACCESS  # Warm CALL to same contract (100)
+        # Inside ERC20 balanceOf
+        + gas_costs.G_VERY_LOW  # PUSH4 selector (3)
+        + gas_costs.G_BASE  # EQ selector match (2)
+        + gas_costs.G_MID  # JUMPI to function (8)
+        + gas_costs.G_JUMPDEST  # JUMPDEST at function start (1)
+        + gas_costs.G_VERY_LOW * 2  # CALLDATALOAD arg (3*2)
+        + gas_costs.G_KECCAK_256  # keccak256 static (30)
+        + gas_costs.G_KECCAK_256_WORD * 2  # keccak256 dynamic for 64 bytes (2*6)
+        + gas_costs.G_COLD_SLOAD  # Cold SLOAD (2100)
+        + gas_costs.G_VERY_LOW * 3  # MSTORE result + RETURN setup (3*3)
+        # RETURN costs 0 gas
+    )
+
+    num_contracts = len(address_stubs.root)
```

Comment:
I think we have to do the pattern (`erc20_contract_*`) discrimination here to have a proper number.

Reply:
For example, if we run execute for all tests, and therefore pass stub contracts that are needed for other tests (like `xen_contract` for example), these other contracts are going to be included in `address_stubs` unconditionally, and we are going to try to send a `balanceOf` to those other contracts that are not ERC20 contracts.